### PR TITLE
Propagate aborted state to dependent signals before firing events

### PR DESCRIPTION
**Pull request by maraisr**
#### 107874a6f093539d8b84cffb0454e954c4e7dee6
<pre>
Propagate aborted state to dependent signals before firing events
<a href="https://bugs.webkit.org/show_bug.cgi?id=278159">https://bugs.webkit.org/show_bug.cgi?id=278159</a>

Reviewed by Chris Dumez.

The implementation and spec assert that dependent signals are aborted
if any of their sources are aborted. However, the intermediate states do not
reflect this during the abort process. Since we iterate through all the dependent
signals, updating the state and then firing the event.

An updated spec is about to land to address this by first propagating
the abort state to any dependent signals, and only then run the
abort steps (run algorithms and fire events).

Worthy notes:
- Dependent signals do not themselves have dependent signals, which
  means it&apos;s unnecessary to call &quot;signal abort recursively&quot;.
- This approach retains the existing event dispatch order while
  ensuring the abort state is synced before any JS runs.

DOM spec PR: <a href="https://github.com/whatwg/dom/pull/1295">https://github.com/whatwg/dom/pull/1295</a>

* LayoutTests/imported/w3c/web-platform-tests/dom/abort/abort-signal-any.any-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/abort/abort-signal-any.any.worker-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/abort/resources/abort-signal-any-tests.js:
  Ported wpt tests from Chromium&apos;s PR addressing this same concern. The
  tests:
  - `Dependent signals for ${desc} are marked aborted before abort events fire ${suffix}` and
  - `Dependent signals for ${desc} are aborted correctly for reentrant aborts ${suffix}`.
* Source/WebCore/dom/AbortSignal.cpp:
(WebCore::AbortSignal::any):
  `check-webkit-style` did not like the `findIf` with `[whitespace/newline]`.
(WebCore::AbortSignal::signalAbort):
  Collect the dependent signals to abort into a vector, then
  mark those as aborted, then
  run the signal&apos;s abort steps, and finally
  run the dependent signals&apos; abort steps.
(WebCore::AbortSignal::markAborted):
  Added this convenience method to set both the `reason` and then
  `m_aborted` state.
(WebCore::AbortSignal::runAbortSteps):
* Source/WebCore/dom/AbortSignal.h:

Canonical link: <a href="https://commits.webkit.org/282387@main">https://commits.webkit.org/282387@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/887ebeec4aad12c4ff994fc37c96f2c16d6238a4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62987 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42343 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15583 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67008 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13591 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65107 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50030 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13875 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50771 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9378 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66056 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39346 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54551 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31455 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36035 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11915 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12467 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57579 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12214 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68703 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6933 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11840 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58087 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6965 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54611 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58292 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5784 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9499 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38163 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39243 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40354 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38985 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/dom/AbortSignal.cpp [hunk 1]
```diff
@@ -79,7 +79,9 @@ Ref<AbortSignal> AbortSignal::any(ScriptExecutionContext& context, const Vector<
 {
     Ref resultSignal = AbortSignal::create(&context);
 
-    auto abortedSignalIndex = signals.findIf([](auto& signal) { return signal->aborted(); });
+    auto abortedSignalIndex = signals.findIf([](auto& signal) {
```
Comment (maraisr) [inline Source/WebCore/dom/AbortSignal.cpp:82]: Note for reviewer: Did `check-webkit-style` change config? Running `Tools/Scripts/check-webkit-style Source/WebCore/dom/AbortSignal*` failed at this line, so fixed. Arguably unrelated change to this PR.

File: Source/WebCore/dom/AbortSignal.cpp [hunk 2]
```diff
@@ -126,25 +128,50 @@ void AbortSignal::signalAbort(JSC::JSValue reason)
     // 1. If signal's aborted flag is set, then return.
     if (m_aborted)
         return;
-    
+
     // 2. Set signal’s aborted flag.
+    markAborted(reason);
+
+    Vector<Ref<AbortSignal>> dependentSignalsToAbort = { };
```
Comment (cdumez) [inline Source/WebCore/dom/AbortSignal.cpp]: `= { }` is unnecessary.

File: Source/WebCore/dom/AbortSignal.cpp [hunk 3]
```diff
@@ -126,25 +128,50 @@ void AbortSignal::signalAbort(JSC::JSValue reason)
...
+
+    for (Ref dependentSignal : std::exchange(m_dependentSignals, { })) {
+        if (!dependentSignal->aborted()) {
+            dependentSignal->markAborted(reason);
+            dependentSignalsToAbort.append(dependentSignal);
```
Comment (cdumez) [inline Source/WebCore/dom/AbortSignal.cpp]: `WTFMove(dependentSignal)`

**Review by cdumez — APPROVED** [review]
LGTM with nits.

File: Source/WebCore/dom/AbortSignal.cpp (hunk 3 above)
Comment (maraisr) [reply Source/WebCore/dom/AbortSignal.cpp]: Thank you. 🙏 I have pushed the nit suggestions. 

Though for my own education, why did we move here? Is it because the m_dependentSignals is emptied and thus ref is --, so we need something else to hold it?

Comment (cdumez) [reply Source/WebCore/dom/AbortSignal.cpp]: This is a performance optimization, it avoids ref-counting churn. We've moving `dependentSignal` straight into the vector, without increasing the ref-count. We can do do since `dependentSignal` is not used after appending to the vector.

Without moving, appending to the vector would increase `dependentSignal`'s refcount, then the refcount would be decreased when the variables goes out of scope (next loop iteration).
